code/modules/power/turbine.dm

```diff
@@ -21,57 +47,116 @@
 	icon_state = "turbine"
 	anchored = 1
 	density = 1
-	var/obj/machinery/compressor/compressor
+	var/opened = 0
+	var/obj/machinery/power/compressor/compressor
 	var/turf/simulated/outturf
 	var/lastgen
+	var/productivity = 1
 
 /obj/machinery/computer/turbine_computer
-	name = "Gas turbine control computer"
+	name = "gas turbine control computer"
 	desc = "A computer to remotely control a gas turbine"
-	icon = 'icons/obj/computer.dmi'
-	icon_state = "airtunnel0e"
-	circuit = /obj/item/weapon/circuitboard/turbine_control
-	anchored = 1
-	density = 1
-	var/obj/machinery/compressor/compressor
-	var/list/obj/machinery/door/poddoor/doors
+	icon_screen = "turbinecomp"
+	icon_keyboard = "tech_key"
+	circuit = /obj/item/weapon/circuitboard/turbine_computer
+	var/obj/machinery/power/compressor/compressor
 	var/id = 0
-	var/door_status = 0
 
 // the inlet stage of the gas turbine electricity generator
 
-/obj/machinery/compressor/New()
+/obj/machinery/power/compressor/New()
 	..()
+	component_parts = list()
+	component_parts += new /obj/item/weapon/circuitboard/power_compressor(null)
+	component_parts += new /obj/item/weapon/stock_parts/manipulator(null)
+	component_parts += new /obj/item/weapon/stock_parts/manipulator(null)
+	component_parts += new /obj/item/weapon/stock_parts/manipulator(null)
+	component_parts += new /obj/item/weapon/stock_parts/manipulator(null)
+	component_parts += new /obj/item/weapon/stock_parts/manipulator(null)
+	component_parts += new /obj/item/weapon/stock_parts/manipulator(null)
+	component_parts += new /obj/item/stack/cable_coil(null, 5)
+	RefreshParts()
+// The inlet of the compressor is the direction it faces
 
 	gas_contained = new
 	inturf = get_step(src, dir)
 
-	spawn(5)
-		turbine = locate() in get_step(src, get_dir(inturf, src))
-		if(!turbine)
-			stat |= BROKEN
+
+/obj/machinery/power/compressor/initialize()
+	..()
+	locate_machinery()
+	if(!turbine)
+		stat |= BROKEN
 
 
 #define COMPFRICTION 5e5
 #define COMPSTARTERLOAD 2800
 
-/obj/machinery/compressor/process()
-	if(!starter)
+
+// Crucial to make things work!!!!
+// OLD FIX - explanation given down below.
+// /obj/machinery/power/compressor/CanPass(atom/movable/mover, turf/target, height=0)
+// 		return !density
+
+/obj/machinery/power/compressor/locate_machinery()
+	if(turbine)
 		return
-	overlays.Cut()
-	if(stat & BROKEN)
+	turbine = locate() in get_step(src, get_dir(inturf, src))
+	if(turbine)
+		turbine.locate_machinery()
+
+/obj/machinery/power/compressor/RefreshParts()
+	var/E = 0
+	for(var/obj/item/weapon/stock_parts/manipulator/M in component_parts)
+		E += M.rating
+	efficiency = E / 6

```

Comment:
Won't this throw a division by 0 error if it has no manipulators in it for whatever reason?